devito/ir/clusters/analysis.py

```diff
@@ -75,32 +75,48 @@ def callback(self, clusters, prefix):
         # Apply the actual callback
         retval = self._callback(clusters, d, prefix)
 
+        # Normalize retval
+        retval = set(as_tuple(retval))
+
         # Update `self.state`
-        if retval is not None:
+        if retval:
             for c in clusters:
                 properties = self.state.properties.setdefault(c, {})
-                properties.setdefault(d, set()).add(retval)
+                properties.setdefault(d, set()).update(retval)
 
         return clusters
 
 
 class Parallelism(Detector):
 
     """
-    Detect SEQUENTIAL, PARALLEL, and PARALLEL_IF_ATOMIC Dimensions.
+    Detect SEQUENTIAL, PARALLEL, PARALLEL_INDEP and PARALLEL_IF_ATOMIC Dimensions.
+
+    Consider an IterationSpace over `n` Dimensions. Let `(d_1, ..., d_n)` be the
+    distance vector of a dependence. Let `i` be the `i-th` Dimension of the
+    IterationSpace. Then:
+
+        * `i` is PARALLEL_INDEP if all dependences have distance vectors:
+
+            (d_1, ..., d_i) = 0
+
+        * `i` is PARALLEL if all dependences have distance vectors:
+
+            (d_1, ..., d_i) = 0, OR
+            (d_1, ..., d_{i-1}) > 0
+
+        * `i` is PARALLEL_IF_ATOMIC if all dependences have distance vectors:
+
+            (d_1, ..., d_i) = 0, OR
+            (d_1, ..., d_{i-1}) > 0, OR
+            the write is known to be an associative and commutative increment
```

Comment:
the 'write' ?

Reply:
'a = b' is a write, where 'a' is being written

Reply:
Agree on that, just in favor of adding quotes...Like The 'write'

Reply:
ok